File: tools/cogstore/signing.py
"""Ed25519 signing / verification over JCS canonical bytes (protocol §7).

The signing input is the catalog with its own `signature` member removed, then
JCS-canonicalized (see `jcs.canonical`). This matches the frozen test vector.
"""
import base64

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey, Ed25519PublicKey)
from cryptography.exceptions import InvalidSignature

from . import jcs

ALG = "ed25519"


def _body(catalog: dict) -> dict:
    """The catalog without its `signature` member — i.e. the signing input source."""
    return {k: v for k, v in catalog.items() if k != "signature"}


def sign_catalog(catalog: dict, *, seed: bytes, key_id: str) -> dict:
    """Return a copy of `catalog` carrying a `signature` member (protocol §7.2)."""
    body = _body(catalog)
    sig = Ed25519PrivateKey.from_private_bytes(seed).sign(jcs.canonical(body))
    out = dict(body)
    out["signature"] = {"key_id": key_id, "alg": ALG,
                        "sig": base64.b64encode(sig).decode()}
    return out


def verify_catalog(catalog: dict, trusted: dict) -> str:
    """Verify the catalog signature; return the key_id used.

    `trusted` maps key_id -> base64 raw 32-byte Ed25519 public key. Raises
    cryptography.exceptions.InvalidSignature on any failure.
    """
    sigblk = catalog.get("signature")
    if not sigblk:
        raise InvalidSignature("no signature member")
    if sigblk.get("alg") != ALG:
        raise InvalidSignature(f"unexpected alg {sigblk.get('alg')!r}")
    kid = sigblk.get("key_id")
    if kid not in trusted:
        raise InvalidSignature(f"untrusted key_id {kid!r}")
    pub = Ed25519PublicKey.from_public_bytes(base64.b64decode(trusted[kid]))
    pub.verify(base64.b64decode(sigblk["sig"]), jcs.canonical(_body(catalog)))
    return kid


def public_key_b64(seed: bytes) -> str:
    """Base64 of the raw 32-byte public key for an Ed25519 seed."""
    raw = Ed25519PrivateKey.from_private_bytes(seed).public_key().public_bytes_raw()
    return base64.b64encode(raw).decode()

File: crates/gearbox/src/signing.rs
//! Ed25519 sign / verify of a catalog over its JCS canonicalization (protocol §7).
//!
//! Producer-independent counterpart to the device-side verifier (seed B4) and the Python
//! `tools/` signer: same algorithm, same test vector. The signing input is the catalog
//! with its own `signature` member removed, then JCS-canonicalized.

use std::collections::HashMap;

use base64::{engine::general_purpose::STANDARD, Engine};
use ed25519_dalek::{Signature, Signer, SigningKey, VerifyingKey};
use serde_json::Value;

use crate::jcs;

/// Trusted keys: `key_id -> raw 32-byte Ed25519 public key`.
pub type TrustStore = HashMap<String, [u8; 32]>;

fn signing_input(catalog: &Value) -> Result<Vec<u8>, String> {
    let mut body = catalog.clone();
    body.as_object_mut()
        .ok_or("catalog is not a JSON object")?
        .remove("signature");
    jcs::canonical(&body)
}

/// Return a copy of `catalog` carrying a `signature` member (protocol §7.2).
pub fn sign_catalog(catalog: &Value, seed: &[u8; 32], key_id: &str) -> Result<Value, String> {
    let msg = signing_input(catalog)?;
    let sig = SigningKey::from_bytes(seed).sign(&msg);
    let mut out = catalog.clone();
    let obj = out.as_object_mut().ok_or("catalog is not a JSON object")?;
    obj.remove("signature");
    obj.insert(
        "signature".to_string(),
        serde_json::json!({
            "key_id": key_id,
            "alg": "ed25519",
            "sig": STANDARD.encode(sig.to_bytes()),
        }),
    );
    Ok(out)
}

/// Verify a catalog's signature; on success return the `key_id` that signed it.
pub fn verify_catalog(catalog: &Value, trust: &TrustStore) -> Result<String, String> {
    let obj = catalog.as_object().ok_or("catalog is not a JSON object")?;
    let sig = obj.get("signature").ok_or("catalog has no signature")?;
    let key_id = sig.get("key_id").and_then(Value::as_str).ok_or("malformed signature.key_id")?;
    let alg = sig.get("alg").and_then(Value::as_str).ok_or("malformed signature.alg")?;
    let sig_b64 = sig.get("sig").and_then(Value::as_str).ok_or("malformed signature.sig")?;

    if alg != "ed25519" {
        return Err(format!("unsupported alg {alg:?}"));
    }
    let pk = trust.get(key_id).ok_or_else(|| format!("untrusted key_id {key_id:?}"))?;
    let vk = VerifyingKey::from_bytes(pk).map_err(|e| format!("bad public key: {e}"))?;

    let msg = signing_input(catalog)?;
    let raw = STANDARD.decode(sig_b64).map_err(|e| format!("bad base64: {e}"))?;
    let signature = Signature::from_slice(&raw).map_err(|e| format!("bad signature: {e}"))?;
    vk.verify_strict(&msg, &signature)
        .map_err(|_| "signature did not verify".to_string())?;
    Ok(key_id.to_string())
}
